interop/reusability.md

```diff
@@ -0,0 +1,1088 @@
+2. *thunk.[[Expression]]* = Expression
+3. Return *thunk*
+
+---
+
+#### The Thunk Exotic Object
+
+Internal Slots:
+
+| Internal Slot    | Type        | Description |
+| ---------------- | ----------- | ----------- |
+| `[[Expression]]` | Expression  | An unevaluated expression |
+| `[[Value]]`      | any         | The value of the JavaScript reference, once evaluated |
+
+##### [[Call]]
+
+1. If `this.[[Value]]` is populated, return `this.[[Value]]`
+2. Let *value* = `GetValue(Evaluate(this.[[Expression]]))?`
+3. `this.[[Value]]` = *value*
+4. Return *value*
+
+---
+
+### Pseudo-Code
+
+Described as pseudo-TypeScript, reifying unevaluated expressions into JavaScript values.
+
+```ts
+function PropertyDefinitionEvaluation(object: Object, decorators: Decorator[], definition: PropertyDefinition) {
+  let propertyNameThunk = new Thunk(definition.PropertyName);
+  let expressionThunk = new Thunk(definition.AssignmentExpression);
+
+  let initialDescriptor = {
+    type: 'property',
+    enumerable: true,
+    configurable: true,
+    writable: true,
+    property: {
+      name: propertyNameThunk,
+      initializer: expressionThunk,
+      get(obj) { return obj[Call(propertyNameThunk)]; },
+      set(obj, val) { obj[Call(propertyNameThunk)] = val; }
+    }
+  };
+
+  let descriptor = decorators.reverse().reduce((descriptor, decorator) => {
+    let possibleDescriptor = Evaluate(decorator)(object, descriptor);
+    return possibleDescriptor === undefined ? descriptor : possibleDescriptor;
+  }, initialDescriptor);
+
+  let { enumerable, configurable, writable, property } = descriptor;
+  let name = Call(property.name);
+  let initializer = property.initializer;
+  
+  if (initializer === null) return;
+  
+  if (typeof 'initializer' === 'function') {
+    let value = Call(initializer);
+    Object.defineProperty(obj, name, { value, enumerable, configurable, writable });
+  } else if (typeof initializer === 'object') {
+    let { get, set } = property;
+    Object.defineProperty(obj, name, { get, set, enumerable, configurable, writable });
+  } else {
+    throw new TypeError(); 
+  }
+}
+
+const EMPTY_SENTINEL = function() {};
+
+function Thunk {
+  let value = EMPTY_SENTINEL;
+  return function() {
+    if (value === EMPTY_SENTINEL) value = Evaluate(expression);
+    return value;
+  }
+}
+```
+
+---
+
+## Appendix: Decorator Descriptor List
+
+##### Decorator Descriptor Interfaces
+
+The shared interface for all decorator descriptors:
+
+```ts
+interface DecoratorDescriptor<Class> {
+  type: string,             // 'property' | 'field' | 'method' | 'accessor'
+  configurable: boolean,    // default: true
+  enumerable: boolean,      // default: true except for methods
+  property: Property<Class>
+}
+```
+
+---
+
+```ts
+interface PropertyDecoratorDescriptor<Class> extends DecoratorDescriptor<Class> {
+  type: string,             // 'property'
+  hint: string,             // 'shorthand' or 'explicit' or 'static'
```

Comment:
What happens if it's a `PropertyDecoratorDescriptor` that is an accessor? It appears you do not receive any hints about which ones it contains like `AccessorDecoratorDescriptor`'s `"getter" | "setter" | "both"`, and `desc.type !==  "accessor"`. Seems inconsistent and confusing.

```js
const obj = {
  @foo
  get bar() {}
};
```